atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/ReadValidationCommitTimestampLoader.java

```diff
@@ -0,0 +1,141 @@
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.impl;
+
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.palantir.atlasdb.futures.AtlasFutures;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.transaction.api.CommitTimestampLoader;
+import com.palantir.atlasdb.transaction.api.TransactionSerializableConflictException;
+import com.palantir.atlasdb.transaction.impl.SerializableTransaction.PartitionedTimestamps;
+import com.palantir.atlasdb.transaction.impl.metrics.TransactionOutcomeMetrics;
+import javax.annotation.Nullable;
+import org.eclipse.collections.api.LongIterable;
+import org.eclipse.collections.api.factory.primitive.LongLongMaps;
+import org.eclipse.collections.api.factory.primitive.LongSets;
+import org.eclipse.collections.api.map.primitive.LongLongMap;
+import org.eclipse.collections.api.map.primitive.MutableLongLongMap;
+import org.eclipse.collections.api.set.primitive.LongSet;
+import org.eclipse.collections.api.set.primitive.MutableLongSet;
+
+/**
+ * Loads commit timestamps for read validation, considering a simulated state of the world where the serializable
+ * transaction that we're performing validation for has already committed.
+ */
+public final class ReadValidationCommitTimestampLoader implements CommitTimestampLoader {
+    private final CommitTimestampLoader delegate;
+    private final long startTs;
+    private final long commitTs;
+    private final TransactionOutcomeMetrics transactionOutcomeMetrics;
+
+    public ReadValidationCommitTimestampLoader(
+            CommitTimestampLoader delegate,
+            long startTs,
+            long commitTs,
+            TransactionOutcomeMetrics transactionOutcomeMetrics) {
+        this.delegate = delegate;
+        this.startTs = startTs;
+        this.commitTs = commitTs;
+        this.transactionOutcomeMetrics = transactionOutcomeMetrics;
+    }
+
+    @Override
+    public ListenableFuture<LongLongMap> getCommitTimestamps(
+            @Nullable TableReference tableRef, LongIterable startTimestamps, boolean shouldWaitForCommitterToComplete) {
+        PartitionedTimestamps partitionedTimestamps = splitTransactionBeforeAndAfter(startTs, startTimestamps);
+
+        ListenableFuture<LongLongMap> postStartCommitTimestamps =
+                getCommitTimestampsForTransactionsStartedAfterMe(tableRef, partitionedTimestamps.afterStart());
+
+        // We are ok to block here because if there is a cycle of transactions that could result in a deadlock,
+        // then at least one of them will be in the ab
+        ListenableFuture<LongLongMap> preStartCommitTimestamps;
+        if (partitionedTimestamps.beforeStart().isEmpty()) {
+            preStartCommitTimestamps = Futures.immediateFuture(LongLongMaps.immutable.empty());
```

Comment:
Could memoize if this happens a bunch?

Reply:
Fair enough - left a comment (preferring not to change semantics for now)